ibis/backends/tests/test_expr_caching.py

```diff
@@ -0,0 +1,146 @@
+from __future__ import annotations
+
+import sys
+
+import pytest
+from pytest import mark
+
+import ibis
+import ibis.common.exceptions as com
+
+pa = pytest.importorskip("pyarrow")
+ds = pytest.importorskip("pyarrow.dataset")
+
+
+@mark.notimpl(["datafusion", "flink", "impala", "trino", "druid"])
+@mark.never(
+    ["mssql"],
+    reason="mssql supports support temporary tables through naming conventions",
+)
+@mark.notimpl(["exasol"], reason="Exasol does not support temporary tables")
+@pytest.mark.never(
+    ["risingwave"],
+    raises=com.UnsupportedOperationError,
+    reason="Feature is not yet implemented: CREATE TEMPORARY TABLE",
+)
+def test_persist_expression(backend, alltypes):
+    non_persisted_table = alltypes.mutate(
+        test_column=ibis.literal("calculation"), other_calc=ibis.literal("xyz")
+    )
+    persisted_table = non_persisted_table.cache()
+    backend.assert_frame_equal(
+        non_persisted_table.to_pandas(), persisted_table.to_pandas()
+    )
+
+
+@mark.notimpl(["datafusion", "flink", "impala", "trino", "druid"])
+@mark.never(
+    ["mssql"],
+    reason="mssql supports support temporary tables through naming conventions",
+)
+@mark.notimpl(["exasol"], reason="Exasol does not support temporary tables")
+@pytest.mark.never(
+    ["risingwave"],
+    raises=com.UnsupportedOperationError,
+    reason="Feature is not yet implemented: CREATE TEMPORARY TABLE",
+)
+def test_persist_expression_contextmanager(backend, con, alltypes):
+    non_cached_table = alltypes.mutate(
+        test_column=ibis.literal("calculation"), other_column=ibis.literal("big calc")
+    )
+    with non_cached_table.cache() as cached_table:
+        backend.assert_frame_equal(
+            non_cached_table.to_pandas(), cached_table.to_pandas()
+        )
+    assert non_cached_table.op() not in con._query_cache.cache
+
+
+@mark.notimpl(["datafusion", "flink", "impala", "trino", "druid"])
+@mark.never(
+    ["mssql"],
+    reason="mssql supports support temporary tables through naming conventions",
+)
+@pytest.mark.never(
+    ["risingwave"],
+    raises=com.UnsupportedOperationError,
+    reason="Feature is not yet implemented: CREATE TEMPORARY TABLE",
+)
+@mark.notimpl(["exasol"], reason="Exasol does not support temporary tables")
+def test_persist_expression_multiple_refs(backend, con, alltypes):
+    non_cached_table = alltypes.mutate(
+        test_column=ibis.literal("calculation"), other_column=ibis.literal("big calc 2")
+    )
+    op = non_cached_table.op()
+    cached_table = non_cached_table.cache()
+
+    cached_df = cached_table.to_pandas()
+    backend.assert_frame_equal(non_cached_table.to_pandas(), cached_df)
+    assert sys.getrefcount(op) == 2
```

Comment:
cc @jcrist for all things reference count related 😬 

This assertion passes, but if you apply this diff (inlining the `cached_table.to_pandas()` call, then refcount assertion fails. Here's the diff:

```patch
diff --git a/ibis/backends/tests/test_expr_caching.py b/ibis/backends/tests/test_expr_caching.py
index 9551f4133..f38df40c4 100644
--- a/ibis/backends/tests/test_expr_caching.py
+++ b/ibis/backends/tests/test_expr_caching.py
@@ -73,8 +73,7 @@ def test_persist_expression_multiple_refs(backend, con, alltypes):
     op = non_cached_table.op()
     cached_table = non_cached_table.cache()
 
-    cached_df = cached_table.to_pandas()
-    backend.assert_frame_equal(non_cached_table.to_pandas(), cached_df)
+    backend.assert_frame_equal(non_cached_table.to_pandas(), cached_table.to_pandas())
     assert sys.getrefcount(op) == 2
 
     name = cached_table.op().name
```

Digging around a bit with `gc.get_referrers`, it looks like a stack frame from inside pyspark's `pandas/conversion.py` is holding a reference to `cached_table` in the inlined case but not when the DataFrame is lifted out into a variable.

Would appreciate any thoughts you have here. I'll probably keep the original code and xfail pyspark for now if there's no obvious way forward here.

Reply:
I'm pushing some other commits, and I'll avoid force pushing for now. You'll want to poke around at https://github.com/ibis-project/ibis/pull/9617/commits/6d0971b73314df22972df194b4168ae817dfc652.

Reply:
Going to force push to pick up the flink fix.